lib/ecto/schema.ex

```diff
@@ -2041,12 +2041,18 @@ defmodule Ecto.Schema do
                            "(no time zone information) or :utc_datetime (time zone is set to UTC)"
   end
 
-  defp check_field_type!(name, type, opts) do
+  defp check_field_type!(name, type, opts, full_type \\ nil) do
+    full_type = if is_nil(full_type), do: type, else: full_type
```

Comment:
Instead of passing it as `nil`, maybe we should change the caller to also pass `type` here and we never have `nil`?